### PR TITLE
Use quinn::read_to_end instead of quinn::read_chunk in a loop

### PR DESCRIPTION
**Pull request by steviez**
#### Summary of Changes
With how we're using [`quinn::read_chunk()`](https://docs.rs/quinn/latest/quinn/struct.RecvStream.html#method.read_chunk) (in a loop until stream is exhausted), we can accomplish the same thing with [`quinn::read_to_end()`](https://docs.rs/quinn/latest/quinn/struct.RecvStream.html#method.read_to_end). Here are the docs on `read_to_end()`:
> Convenience method to read all remaining data into a buffer
> 
> The returned future fails with [ReadToEndError::TooLong](https://docs.rs/quinn/latest/quinn/enum.ReadToEndError.html#variant.TooLong) if it’s longer than size_limit bytes. Uses unordered reads to be more efficient than using AsyncRead would allow. size_limit should be set to limit worst-case memory use.
>
> If unordered reads have already been made, the resulting buffer may have gaps containing arbitrary data.

Most importantly, `read_to_end()` will still use unordered reads. Additionally, having the entire buffer to copy at once means only one memcpy into packet via `copy_from_slice` instead of one-per-chunk.

Lastly, add a timeout on `read_to_end()` (and stat to track occurrences of this) per t-nelson's comments below. Default is 2500 (ms), but operator can alter by setting `SOLANA_QUIC_READ_TIMEOUT_MS` env variable.

### CODE REVIEW

**Comment by t-nelson** [comment]
would this not open us to dos in the case that an attacker sends us less than the cap _and_ holds the stream open?

**Comment by steviez** [comment]
> would this not open us to dos in the case that an attacker sends us less than the cap _and_ holds the stream open?
> 
> EDIT: I guess the chunked read was technically susceptible to the same attack. Something something [timeout](https://docs.rs/tokio/latest/tokio/time/fn.timeout.html)

Good call - any thoughts as to what a reasonable timeout value may be? The connection/stream should be open by this point, but still think we might need to allow several seconds

**Comment by t-nelson** [comment]
> any thoughts as to what a reasonable timeout value may be?

yeah it's a little tricky.  I'm having a hard time seeing this "all protocols sharing the same connection pool" last very long.  it seems to me that trying to balance individual protocol constraints without negative externalities affecting the others will be needlessly complex.  i see this shaking out to per-protocol supervised connection caches with an L2 supervisor to flag abuse.

tl;dr, 2500ms is probably a good start :slightly_smiling_face: 

Maybe consider an envvar that we can tune it with during rollout (and kill ASAP!)

File: streamer/src/quic.rs [hunk 1]
```diff
@@ -29,6 +30,17 @@ use {
     },
 };
 
+// TODO: remove this ENV variable when we have confidence in a good default timeout value
+lazy_static! {
+    static ref QUIC_READ_TIMEOUT_MS: u64 = env::var("SOLANA_QUIC_READ_TIMEOUT_MS")
+        .map(|x| x.parse().unwrap_or(2500))
```
Comment (t-nelson) [inline streamer/src/quic.rs:36]:
```suggestion
        .and_then(|x| x.parse().ok())
```
one magic number is better than two!

File: streamer/src/quic.rs [hunk 2]
```diff
@@ -356,20 +346,23 @@ fn handle_connection(
...
-                                break;
+                        if !stream_exit.load(Ordering::Relaxed) {
+                            let timeout_ms = get_read_timeout_ms();
+                            let timeout_read_to_end = timeout(
+                                Duration::from_millis(timeout_ms),
+                                stream.read_to_end(PACKET_DATA_SIZE),
+                            )
+                            .await;
+                            if let Ok(read_to_end) = timeout_read_to_end {
+                                handle_read_to_end(&read_to_end, &remote_addr, &packet_sender);
```
Comment (t-nelson) [inline streamer/src/quic.rs:360]:
```suggestion
                            if let Ok(read_to_end) = timeout_read_to_end.as_ref() {
                                handle_read_to_end(read_to_end, &remote_addr, &packet_sender);
```

to avoid the destructure

**Review by t-nelson — APPROVED** [review]
lgtm! just nits.  :+1: for metrics

**Comment by sakridge** [comment]
Performance difference?

**Comment by sakridge** [comment]
> Most importantly, read_to_end() will still use unordered reads. Additionally, having the entire buffer to copy at once means only one memcpy into packet via copy_from_slice instead of one-per-chunk.

https://docs.rs/quinn/latest/src/quinn/recv_stream.rs.html#334

If you look at the implementation, it just pushes the chunks onto a Vec and then it copies them out to provide the answer. This gives you another copy that we didn't have before.

**Comment by steviez** [comment]
>

**Comment by steviez** [comment]
> If you look at the implementation, it just pushes the chunks onto a Vec and then it copies them out to provide the answer. This gives you another copy that we didn't have before. (and vector allocation/manipulation of pushing the chunks)

Yep, you're right; didn't dig down deep enough. I'll refactor this PR, as I think we still want the timeout (albeit maybe shorter default given that it will be per-chunk instead of for the whole thing)

**Comment by sakridge** [comment]
> > If you look at the implementation, it just pushes the chunks onto a Vec and then it copies them out to provide the answer. This gives you another copy that we didn't have before. (and vector allocation/manipulation of pushing the chunks)
> 
> Yep, you're right; didn't dig down deep enough. I'll refactor this PR, as I think we still want the timeout (albeit maybe shorter default given that it will be per-chunk instead of for the whole thing)

Did we test that it's actually a problem? I am usually seeing it time-out the connection already which will close the stream automatically if there is no activity.
